File: 347.top-k-frequent-elements.java
/*
 * @lc app=leetcode id=347 lang=java
 *
 * [347] Top K Frequent Elements
 */

// @lc code=start

import java.util.AbstractMap;
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> frequence = new HashMap<>();
        for (int n : nums) {
            frequence.put(n, frequence.getOrDefault(n, 0) + 1);
        }
        PriorityQueue<AbstractMap.SimpleEntry<Integer, Integer>> pq = new PriorityQueue<>(
                (a, b) -> a.getValue() - b.getValue());
        for (Map.Entry<Integer, Integer> entry : frequence.entrySet()) {
            pq.add(new AbstractMap.SimpleEntry<Integer, Integer>(entry.getKey(), entry.getValue()));
            if (pq.size() > k) {
                pq.remove();
            }
        }
        int[] ret = new int[k];
        while (k > 0) {
            ret[--k] = pq.remove().getKey();
        }
        return ret;
    }
}
// @lc code=end

File: 347.top-k-frequent-elements.cpp
/*
 * @lc app=leetcode id=347 lang=cpp
 *
 * [347] Top K Frequent Elements
 */

// @lc code=start
#include <algorithm>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:

    using element_type = pair<int, unsigned>;

    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, unsigned> umap;
        for (const auto &el : nums) ++umap[el];
        auto freq_cmp = [](const element_type &lhs, const element_type &rhs) -> bool {
            return lhs.second > rhs.second;
        };
        vector<element_type> vec { umap.cbegin(), umap.cend() };
        partial_sort(vec.begin(), vec.begin() + k, vec.end(), freq_cmp);
        vector<int> ret;
        transform(vec.cbegin(), vec.cbegin() + k, back_inserter(ret), [](const auto &e) {
            return e.first;
        });
        return ret;
    }
};
// @lc code=end
